### PR TITLE
Update CHANGELOG.md for grpc and proto changes

### PR DESCRIPTION
**Pull request by peter-edge**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,23 @@ v1.9.0-dev (unreleased)
...
+-   x/grpc: Add the `grpcheader` package to make it easier to deal with
+    required x/grpc headers when using native gRPC libraries.
+-   x/grpc: Start work on configuration handling.
+-   x/protobuf: Change the `yarpc-protobuf-raw-response` header key to
+    `rpc-protobuf-raw-response`.
+-   x/protobuf: Change the name of the encoding from `protobuf` to `proto`.
+-   x/protobuf: Handle JSON-encoded protobuf requests and return JSON-encoded
+    protobuf responses if the `rpc-encoding` header is set to `json`, and allow
+    protobuf clients to use JSON by using the `protobuf.UseJSON` option.
+-   x/protobuf: Call `yarpc.RegisterClientBuilder` from code generated by
```
Comment (breerly) [inline CHANGELOG.md]: Internal detail, not needed here

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,6 +20,23 @@ v1.9.0-dev (unreleased)
...
 -   Pool inbound thrift request buffers to reduce allocations.
+-   x/grpc: Use `rpc-caller`, `rpc-service`, `rpc-encoding`, `rpc-shard-key`,
+    `rpc-routing-key`, `rpc-routing-delegate` headers.
+-   x/grpc: Remove `yarpc-grpc-app-` application header prefix.
+-   x/grpc: Require `rpc-caller`, `rpc-service`, `rpc-encoding` headers.
+    `rpc-encoding` may be substituted for `content-type` if `content-type`
+    follows the content subtype structure for gRPC.
+-   x/grpc: Add the `grpcheader` package to make it easier to deal with
+    required x/grpc headers when using native gRPC libraries.
+-   x/grpc: Start work on configuration handling.
```
Comment (breerly) [inline CHANGELOG.md]: Internal detail, not needed here

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,6 +20,23 @@ v1.9.0-dev (unreleased)
...
 -   x/config: Fix bug where Chooser and Updater fields could not be interpolated.
 -   Pool inbound thrift request buffers to reduce allocations.
+-   x/grpc: Use `rpc-caller`, `rpc-service`, `rpc-encoding`, `rpc-shard-key`,
+    `rpc-routing-key`, `rpc-routing-delegate` headers.
+-   x/grpc: Remove `yarpc-grpc-app-` application header prefix.
+-   x/grpc: Require `rpc-caller`, `rpc-service`, `rpc-encoding` headers.
+    `rpc-encoding` may be substituted for `content-type` if `content-type`
+    follows the content subtype structure for gRPC.
+-   x/grpc: Add the `grpcheader` package to make it easier to deal with
+    required x/grpc headers when using native gRPC libraries.
```
Comment (breerly) [inline CHANGELOG.md]: Internal detail, not needed here

File: CHANGELOG.md [hunk 4]
```diff
@@ -20,6 +20,23 @@ v1.9.0-dev (unreleased)
 -   x/config: Fix bug where embedded struct fields could not be interpolated.
 -   x/config: Fix bug where Chooser and Updater fields could not be interpolated.
 -   Pool inbound thrift request buffers to reduce allocations.
+-   x/grpc: Use `rpc-caller`, `rpc-service`, `rpc-encoding`, `rpc-shard-key`,
+    `rpc-routing-key`, `rpc-routing-delegate` headers.
+-   x/grpc: Remove `yarpc-grpc-app-` application header prefix.
+-   x/grpc: Require `rpc-caller`, `rpc-service`, `rpc-encoding` headers.
+    `rpc-encoding` may be substituted for `content-type` if `content-type`
+    follows the content subtype structure for gRPC.
+-   x/grpc: Add the `grpcheader` package to make it easier to deal with
```
Comment (breerly) [inline CHANGELOG.md]: Internal detail, not needed here

File: CHANGELOG.md [hunk 5]
```diff
@@ -20,6 +20,23 @@ v1.9.0-dev (unreleased)
 -   x/config: Fix bug where embedded struct fields could not be interpolated.
 -   x/config: Fix bug where Chooser and Updater fields could not be interpolated.
 -   Pool inbound thrift request buffers to reduce allocations.
+-   x/grpc: Use `rpc-caller`, `rpc-service`, `rpc-encoding`, `rpc-shard-key`,
+    `rpc-routing-key`, `rpc-routing-delegate` headers.
```
Comment (breerly) [inline CHANGELOG.md]: Internal detail, not needed here

File: CHANGELOG.md [hunk 6]
```diff
@@ -20,6 +20,23 @@ v1.9.0-dev (unreleased)
 -   x/config: Fix bug where embedded struct fields could not be interpolated.
 -   x/config: Fix bug where Chooser and Updater fields could not be interpolated.
 -   Pool inbound thrift request buffers to reduce allocations.
+-   x/grpc: Use `rpc-caller`, `rpc-service`, `rpc-encoding`, `rpc-shard-key`,
```
Comment (breerly) [inline CHANGELOG.md]: Internal detail, not needed here

**Comment by peter-edge** [comment]
@breerly I don't think your comments are on the right lines, you have comments "not needed here" twice on the same bullet point in some places, can you make sure the right ones are done?

**Comment by breerly** [comment]
I blame github mobile - anyways, do you the host of my review? Reduce the lines added to only need-tos.

**Comment by breerly** [comment]
get the jist of my review*

**Comment by peter-edge** [comment]
How's this?

File: CHANGELOG.md [hunk 7]
```diff
@@ -15,11 +15,18 @@ v1.9.0-dev (unreleased)
...
 -   x/config: Fix bug where embedded struct fields could not be interpolated.
 -   x/config: Fix bug where Chooser and Updater fields could not be interpolated.
 -   Pool inbound thrift request buffers to reduce allocations.
+-   x/grpc: Remove NewInbound and NewSingleOutbound in favor of
+    functions on Transport
+-   x/grpc: Use `rpc-caller`, `rpc-service`, `rpc-encoding`, `rpc-shard-key`,
+    `rpc-routing-key`, `rpc-routing-delegate` headers.
+-   x/protobuf: Handle JSON-encoded protobuf requests and return JSON-encoded
+    protobuf responses if the `rpc-encoding` header is set to `json`, and allow
+    protobuf clients to use JSON by using the `protobuf.UseJSON` option.
```
Comment (abhinav) [inline CHANGELOG.md]: How about,

> [..] if the `rpc-encoding` header is set to `json`. Protobuf clients may use JSON by supplying the `protobuf.UseJSON` option.

File: CHANGELOG.md [hunk 8]
```diff
@@ -15,11 +15,18 @@ v1.9.0-dev (unreleased)
...
+-   x/protobuf: Call `yarpc.RegisterClientBuilder` from code generated by
+    `protoc-gen-yarpc-go`.
```
Comment (abhinav) [inline CHANGELOG.md]: The changelog should reflect how the change affects the user.

> x/protobuf: Support instantiating clients with `yarpc.InjectClients`.

File: CHANGELOG.md (hunk 7 above)
Comment (peter-edge) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 8 above)
Comment (peter-edge) [inline CHANGELOG.md]: done

**Comment by peter-edge** [comment]
Documented - the `UseJSON` tag has godoc documentation I think, the `proto:"json"` isn't explicitly documented anywhere I don't think, but I don't know if this is documented for thrift either. Note that this feature isn't really fully-baked yet, and this is still in x/.

**Comment by peter-edge** [comment]
@breerly @abhinav merge or change this to whatever you want if it needs more work before you do the release, I think all comments are addressed

File: CHANGELOG.md [hunk 9]
```diff
@@ -15,11 +15,17 @@ v1.9.0-dev (unreleased)
...
 -   x/config: Fix bug where Chooser and Updater fields could not be interpolated.
 -   Pool inbound thrift request buffers to reduce allocations.
+-   x/grpc: Remove NewInbound and NewSingleOutbound in favor of
+    functions on Transport.
+-   x/grpc: Use `rpc-caller`, `rpc-service`, `rpc-encoding`, `rpc-shard-key`,
+    `rpc-routing-key`, `rpc-routing-delegate` headers.
+-   x/protobuf: Handle JSON-encoded protobuf requests and return JSON-encoded
+    protobuf responses if the `rpc-encoding` header is set to `json`. Protobuf
+    clients may use JSON by supplying the `protobuf.UseJSON` option.
+-   x/protobuf: Support instantiating clients with yarpc.InjectClients.
```
Comment (willhug) [inline CHANGELOG.md]: change protobuf encoding header from "protobuf" to "proto"

**Comment by peter-edge** [comment]
I just deleted that from a comment Grayson made, fight it out with Grayson,
I'm not changing it back

**Comment by willhug** [comment]
wut, but changing the encoding type from "protobuf" to "proto" is a wire compatibility change...

**Comment by peter-edge** [comment]
Fight it out with Grayson, I consider this PR (way over)done

**Comment by abhinav** [comment]
> [..] the proto:"json" isn't explicitly documented anywhere I don't think, but I don't know if this is documented for thrift either. [..]

It's documented: https://godoc.org/go.uber.org/yarpc/encoding/thrift#hdr-Automatically_Building_Clients. To be clear, I don't expect this to be changed in this PR, but it's definitely something that should be documented if you're advertising the `UseJSON` feature, especially because scaffolded YARPC services use InjectClients by default.

**Comment by peter-edge** [comment]
Ok will do in the future, this PR has gotten out of control for what should
have been a quick fix and merge

**Pull request merged**